### PR TITLE
[UTOPIA-840]add validation message to mandatory field

### PR DESCRIPTION
**Pull request by binggaofw**
# Description

This PR includes the following proposed change(s):

- add label to mandatory field using red color
- dispaly  warning message in mandatory field
Screen Shot 2023-03-24 at 4 12 49 PM

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Development Dependency Working Agreement
- [x] My code DOES NOT include the importing of new dependencies into the DPIA ecosystem
- [ ] My code DOES include the importing of new dependencies into the DPIA ecosystem
**If new dependencies are being introduced to the DPIA ecosystem:**
- [ ] The functionality of the dependency drastically reduces code complexity and makes my changes more easily maintainable and readible 
- [ ] The dependency being introduced does not contain multiple layers of nested dependencies introducing maintainability complexity to the DPIA ecosystem

## Frontend Development Changes
- [ ] N/A
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to project documentation or diagrams that reflect my changes
- [ ] New and existing unit tests pass locally with my changes
- [ ] My code follows Airbnb React Style Guidelines

## API Development Changes
- [x] N/A
- [ ] I have performed a self-review of my own code
- [ ] My code follows standards and practices outlined in the BC Government API Development Guidelines
- [ ] New and existing unit tests pass locally with my changes
- [ ] My changes includes Swagger documentation updates that reflect the changes I am introducing

## Definition of Done

Definition of Done

### CODE REVIEW

**Comment by binggaofw** [comment]
display all the error message 
Screen Shot 2023-03-27 at 8 53 59 AM

File: src/frontend/src/components/public/PIAFormTabs/intake/index.tsx [hunk 1]
```diff
@@ -67,30 +82,43 @@ export const PIAFormIntake = () => {
             <div className="row">
               <InputText
                 label="Title"
+                fieldMandatoryLabel="(required)"
```
Comment (kushal-arora-fw) [inline src/frontend/src/components/public/PIAFormTabs/intake/index.tsx:85]: isn't "required" label consistent for all fields? So, why send from all InputText usages?

File: src/frontend/src/components/public/PIAFormTabs/intake/index.tsx [hunk 2]
```diff
@@ -67,30 +82,43 @@ export const PIAFormIntake = () => {
...
+                <Dropdown
+                  id="ministry-select"
+                  fieldMandatoryLabel="(required)"
+                  value={intakeForm?.ministry || ''}
+                  label="Ministry"
+                  options={MinistryList}
+                  changeHandler={(e) =>
+                    stateChangeHandler(e.target.value, 'ministry')
+                  }
+                  required={true}
```
Comment (kushal-arora-fw) [inline src/frontend/src/components/public/PIAFormTabs/intake/index.tsx:105]: can we not use this prop, `required`? If required is true, then show the label?

File: src/frontend/src/pages/PIAForm/index.tsx [hunk 3]
```diff
@@ -476,15 +487,30 @@ const PIAFormPage = () => {
     if (!pia?.title) {
       invalid = true;
       formId = 'title';
+      setValidationMessages((prevState) => ({
+        ...prevState,
+        piaTitle: 'Error: Please enter a title.',
+      }));
     }
-    if (!pia?.ministry && !invalid) {
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIAForm/index.tsx:480]: why removing !invalid?

I will let @antsand  - give this a look.

File: src/frontend/src/pages/PIAForm/index.tsx [hunk 4]
```diff
@@ -544,6 +578,12 @@ const PIAFormPage = () => {
       }
       event.preventDefault();
       event.stopPropagation();
+
+      navigate(
+        buildDynamicPath(routes.PIA_INTAKE_EDIT, {
+          id: pia?.id,
+        }),
+      );
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIAForm/index.tsx:586]: why navigate if the form is invalid?

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 3 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:480]: I talk to Jordan, She mentioned we should show all the error message at the same time if the user click the submit button

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 4 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:586]: it is in AC. we need to route the user to the error place

File: src/frontend/src/components/common/Dropdown/index.tsx [hunk 5]
```diff
@@ -12,7 +13,11 @@ const Dropdown = ({
 }: IDropdown) => {
   return (
     <div className={`form-group ${optionalClass}`}>
-      {label && <label htmlFor={id}>{label}</label>}
+      {label && (
+        <label htmlFor={id}>
+          {label} <span className="text-danger"> {fieldMandatoryLabel}</span>
```
[the remark this answers is not in the document]
Comment (binggaofw) [reply src/frontend/src/components/common/Dropdown/index.tsx:18]: good catch, they are not the same, we have our own error color, i will change it

File: src/frontend/src/components/public/PIAFormTabs/intake/index.tsx (hunk 2 above)
Comment (binggaofw) [reply src/frontend/src/components/public/PIAFormTabs/intake/index.tsx:103]: yes, refactor follow your advice, it is much clear now

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 4 above)
Comment (antsand) [reply src/frontend/src/pages/PIAForm/index.tsx:586]: Don't we need to scroll to the error? Why navigate? Line 576 will do it for you.

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 3 above)
Comment (antsand) [reply src/frontend/src/pages/PIAForm/index.tsx:480]: There might be an validation abstraction tool to do this in 'Mighty react'.  Is there not? I wrote it in raw JS as it was easier for me to compute.

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 4 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:586]: this is for view mode redirect 
Screenshot 2023-03-28 at 8 43 53 AM

File: src/frontend/src/components/common/InputText/InputText.tsx [hunk 6]
```diff
@@ -8,6 +8,7 @@ type SupportedInputTypes = 'text' | 'email';
 interface InputTextProps {
   id?: string;
   label?: string;
+  fieldMandatoryLabel?: string;
```
Comment (kushal-arora-fw) [inline src/frontend/src/components/common/InputText/InputText.tsx:11]: Remove this unused prop

File: src/frontend/src/components/common/InputText/InputText.tsx [hunk 7]
```diff
@@ -26,6 +27,7 @@ interface InputTextProps {
 const InputText = ({
   id = '',
   label = '',
+  fieldMandatoryLabel = '',
```
Comment (kushal-arora-fw) [inline src/frontend/src/components/common/InputText/InputText.tsx:30]: Remove this unused prop

File: src/frontend/src/pages/PIAForm/index.tsx [hunk 8]
```diff
@@ -27,6 +27,13 @@ export type PiaStateChangeHandlerType = (
 
 export type PiaFormOpenMode = 'edit' | 'view';
 
+export interface PiaValidationMessage {
+  piaTitle?: string | null;
+  piaMinistry?: string | null;
+  piaBranch?: string | null;
+  piaInitialDescription?: string | null;
+}
+
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIAForm/index.tsx:36]: shouldn't these validation messages be inside intake form component and not the parent component?

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 4 above)
Comment (antsand) [reply src/frontend/src/pages/PIAForm/index.tsx:586]: let us clarify with Jordan about this. This seems a bit convoluted.

**Comment by kushal-arora-fw** [comment]
- No banner message
- Banner message and disabled submit should only be for view mode


image

File: src/frontend/src/components/common/Alert/_alert.scss [hunk 9]
```diff
@@ -38,12 +37,12 @@
   }
 
   .alert-close {
-    position: relative;
```
Comment (antsand) [inline src/frontend/src/components/common/Alert/_alert.scss:41]: Are you sure this is right? You might break alert for other alert messages.

File: src/frontend/src/pages/PIAForm/index.tsx [hunk 10]
```diff
@@ -476,64 +506,35 @@ const PIAFormPage = () => {
...
+    if (!pia?.initiativeDescription) {
       invalid = true;
       formId = 'initiativeDescription';
-    }
-    if (!pia?.initiativeScope && !invalid) {
-      invalid = true;
-      formId = 'initiativeScope';
-    }
-    if (!pia?.dataElementsInvolved && !invalid) {
```
Comment (antsand) [inline src/frontend/src/pages/PIAForm/index.tsx:528]: should we not validate this?

File: src/frontend/src/pages/PIAForm/index.tsx [hunk 11]
```diff
@@ -476,64 +506,35 @@ const PIAFormPage = () => {
...
-      invalid = true;
-      formId = 'dataElementsInvolved';
-    }
-    if (pia?.hasAddedPiToDataElements === false) {
```
Comment (antsand) [inline src/frontend/src/pages/PIAForm/index.tsx:532]: And also validate this if 'no pi' is selected?

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 10 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:528]: this value is a radio button, I think we are giving it a default value

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 11 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:532]: good catch, just ask UI designer for clarification

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 10 above)
Comment (antsand) [inline src/frontend/src/pages/PIAForm/index.tsx:528]: I do believe this should still be validated. As by default we need either a 'yes' or a 'no' or 'null'. But there can be an instance where non of the elements are checked (someone trying to right click go to inspect element etc .. and send nothing checked) the api will give an error. The frontend should not allow this to happen.  MY 2 cents.

Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:528]: agree, add back

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 8 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:36]: will create a refactor ticket for this

File: src/frontend/src/pages/PIAForm/index.tsx (hunk 3 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIAForm/index.tsx:480]: will create a refactor ticket for this

**Comment by binggaofw** [comment]
> * No banner message
> * Banner message and disabled submit should only be for view mode
> 
> image

I update the message bootstrap class to make sure all the time to show the message and only show in view mode

**Pull request merged**